passw0rd/src/main/kotlin/com/virgilsecurity/passw0rd/protocol/Protocol.kt

```diff
@@ -0,0 +1,198 @@
+     *
+     * @throws IllegalArgumentException
+     * @throws ProtocolException
+     * @throws PheException
+     */
+    @Throws(IllegalArgumentException::class, ProtocolException::class, PheException::class)
+    fun enrollAccount(password: String): CompletableFuture<EnrollResult> = GlobalScope.async {
+        if (password.isBlank()) Utils.shouldNotBeEmpty("password")
+
+        Passw0rdProtos.EnrollmentRequest.newBuilder().setVersion(currentVersion).build().run {
+            httpClient.firePost(
+                    this,
+                    HttpClientProtobuf.AvailableRequests.ENROLL,
+                    authToken = appToken,
+                    responseParser = Passw0rdProtos.EnrollmentResponse.parser()
+            ).let { response ->
+                val enrollResult = try {
+                    pheClients[response.version]!!.enrollAccount(response.response.toByteArray(),
+                                                                 password.toByteArray())
+                } catch (exception: PheException) {
+                    throw InvalidProofException()
+                }
+
+                val enrollmentRecord = Passw0rdProtos.DatabaseRecord
+                        .newBuilder()
+                        .setVersion(currentVersion)
+                        .setRecord(ByteString.copyFrom(enrollResult.enrollmentRecord))
+                        .build()
+                        .toByteArray()
+
+                EnrollResult(enrollmentRecord, enrollResult.accountKey)
+            }
+        }
+    }.asCompletableFuture()
+
+    /**
+     * This function verifies a [password] against [enrollmentRecord] using passw0rd service.
+     *
+     * @throws IllegalArgumentException
+     * @throws ProtocolException
+     * @throws PheException
+     * @throws InvalidPasswordException
+     * @throws InvalidProtobufTypeException
+     */
+    @Throws(IllegalArgumentException::class,
+            ProtocolException::class,
+            PheException::class,
+            InvalidPasswordException::class,
+            InvalidProtobufTypeException::class)
+    fun verifyPassword(password: String, enrollmentRecord: ByteArray): CompletableFuture<ByteArray> = GlobalScope.async {
+        if (password.isBlank()) Utils.shouldNotBeEmpty("password")
+        if (enrollmentRecord.isEmpty()) Utils.shouldNotBeEmpty("enrollmentRecord")
+
+        val (version, record) = try {
+            Passw0rdProtos.DatabaseRecord.parseFrom(enrollmentRecord).let {
+                it.version to it.record.toByteArray()
+            }
+        } catch (e: InvalidProtocolBufferException) {
+            throw InvalidProtobufTypeException()
+        }
+
+        if (pheClients[version] == null)
+            throw NoKeysFoundException("Unable to find keys corresponding to record's version $version.")
+
+        val request = pheClients[version]!!.createVerifyPasswordRequest(password.toByteArray(), record)
```

Comment:
Use of force unwrap is not recommended

Reply:
Try to store pheClients[version] in separate variable and check it for null

Reply:
good point